### PR TITLE
Replace skipped CO unit tests and collect the dead upk suite

### PR DESCRIPTION
**Pull request by david-j-stoll**
## Context & Motivation

- Fixes the backlog item covering skipped backend unit tests: "All skipped tests are either fixed or removed"
- Related PR: [#1199](https://github.com/MyFriendBen/benefits-api/pull/1199), which stood up the CI workflow and introduced the skips
- A live `fps` income-limit bug was found while doing this work. It is filed separately and deliberately **not** fixed here — see Notes for Reviewers.

When CI was first stood up in `5249ade2`, nine Colorado calculator test files were skipped wholesale rather than fixed:

```
@skip("Calculator constructor signature changed - needs API update")
```

That marker undersells it. Those tests were written against a calculator API two generations old — `Calculator(screen)` (or `Calculator(screen, prev_eligibility_list)`) returning an `.eligibility` **dict** — versus today's `Calculator(screen, program, data, missing_dependencies)` with `.eligible() -> Eligibility`. They were rewrites, not signature tweaks, which is why they sat skipped for ten months.

The practical impact: **nine calculators, seven of them active in Colorado, had zero unit test coverage.** Validations are deprecated, so nothing else covered them either.

## Changes Made

- **Replaced all 9 skipped test files with mock-based unit tests** in the current idiom (matching `wa/head_start`, `ks/k40h`). Each file covers class attributes and registry membership, household-level conditions with boundary cases, member-level conditions with boundary cases, full `eligible()` composition, `value()`, and `can_calc()` / `DependencyError`.

  | Program | Abbrev | Tests | Active |
  |---|---|---|---|
  | `co/connect_for_health` | `cfhc` | 47 | ✅ |
  | `co/head_start` | `co_head_start` | 49 | ✅ |
  | `co/property_credit_rebate` | `cpcr` | 45 | ✅ |
  | `co/medicaid/family_planning_services` | `fps` | 44 | ✅ |
  | `co/dental_health_care_seniors` | `cdhcs` | 40 | ✅ |
  | `co/omnisalud` | `omnisalud` | 34 | ✅ |
  | `co/denver_preschool_program` | `dpp` | 34 | ✅ |
  | `co/energy_resource_center` | `erc` | 31 | ❌ inactive |
  | `co/reproductive_health_care` | `rhc` | 28 | ❌ inactive |

  `erc` and `rhc` are inactive today but still registered in `co_calculators`, so they'd run the moment either is switched on. Covered rather than deleted.

- **Rescued a tenth file that was never collected.** `co/universal_preschool/test.py` matched none of the `python_files` patterns in `pytest.ini` (`tests.py`, `test_*.py`, `*_tests.py`) — singular `test.py` matches nothing. Its 8 tests passed but had never run in CI, and there was no `@skip` to make that visible. Moved to `tests/test_universal_preschool.py`.

- **Moved every file into a `tests/` package** (`tests/test_<program>.py`) to match the convention the newer suites use.

**Suite: 2645 collected / 19 skipped → 2986 collected / 0 skipped.** 352 net-new tests. The only skip left in the repo is the intentional `HUD_API_TOKEN` guard in `conftest.py:233`, which exists so fork PRs without secrets don't fail on integration tests — that one should stay.

## Testing

- Migrations to run: none
- Configuration updates needed: none
- Environment variables/settings to add: none
- Manual testing steps:
  ```
  pytest                                    # 2986 passed, 0 skipped
  pytest programs/programs/co -q            # the changed surface
  black --check programs/programs/co        # clean
  ```
  No manual/browser testing applies — this PR is tests only and touches no application code.

## Deployment

- Post-deployment scripts needed: none
- Production config updates: none
- Admin updates needed: none
- Notify team/users of: nothing blocking. Worth flagging to whoever picks up the `fps` bug ticket that the `fps` tests in this PR pin the buggy behaviour on purpose and will need editing as part of that fix — it's written into that ticket's acceptance criteria.

Tests only. No application code, models, migrations, or config are touched, so there is nothing to deploy beyond the merge itself.

## Notes for Reviewers

**No production code changed.** Every calculator behaves exactly as it did before this PR. Where the tests document something that looks wrong, they pin current behaviour and say so in a comment rather than quietly correcting it.

**Four behaviours worth a look, all pre-existing:**

1. **`fps` income limit is ~2× too generous** (the one real bug here, filed separately). `household_eligible` sizes the FPL lookup with `household_size + len(e.eligible_members)`, but `eligible_members` is appended to for *every* member in `ProgramCalculator.eligible()`, not just eligible ones — every other call site in the codebase filters on `.eligible` first. Regression from `090100de` (Sept 2024): before that refactor `household_eligible()` built its own fresh `Eligibility()`, so the count was structurally always `0` and the limit was `2.65 × FPL(household_size)`. For a 2-person household the limit went from $56,047 to $85,197. Not fixed here because the correct target isn't obvious from the code alone — "plain household size" vs. "household size plus genuinely eligible members" is a program question, and it's the first acceptance criterion on that ticket.

2. **Google Sheets caches fail silently, in production too.** `Cache.fetch()` catches exceptions and returns the class default. So if the Head Start counties sheet is unreachable, `CoHeadStart.counties.fetch()` returns `{}` and *every Colorado household becomes ineligible* — no error, no alert, just a program that quietly stops matching anyone. Same mechanism in `ConnectForHealth`, where an empty `county_values` makes `member_value` raise `KeyError` on `values[county]` instead of degrading. Both are characterised in tests here.

   The `ConnectForHealth` half is worse than "that program returns nothing": `program.eligibility()` is wrapped only in `except DependencyError` (`screener/views.py:481`) and there is no outer handler in `eligibility_results`, so that `KeyError` propagates all the way out and 500s the **entire** `/eligibility` response — every program on the screen, not just `cfhc`. Filing a ticket for both unless a reviewer objects.

3. **`PropertyCreditRebate._is_surviving_spouse` is hardcoded `False`** because the screener has no question for it. Pinned with a test so the stub isn't forgotten.

4. **`CoHeadStart` decides on the first county its sheet recognises.** The county loop `break`s on the first match, so for a zipcode spanning several counties a leading county flagged `FALSE` rejects the household even when a later one is flagged `TRUE`. Not an edge case: 359 of the 526 zipcodes in `co_county_zips` map to more than one county (`80022` → Adams + Denver), and the multi-county path is what runs whenever `screen.county` is unset. Whether the rule should be "any participating county in the zip" or "the first one" is a program question, so `TestCoHeadStartMultiCountyZip` pins it as-is rather than changing it. A county *absent* from the sheet does not stop the loop, which is a separate branch and covered too.

**On test style:** these are unit tests — the screen, program, and members are mocked, and no test touches the database for the code under test. Three deliberate exceptions where a real object is more faithful than a mock and still needs no DB: unsaved `Insurance` instances (so the `dont_know` aliasing inside `has_insurance_types` is genuinely exercised rather than mocked away); unsaved `HouseholdMember` instances in the `cpcr` tests (mocking `has_disability` collapsed `disabled`, `visually_impaired`, and `long_term_disability` into one return value, so dropping any one of the three went undetected); and the real `co_county_zips` data (so the zipcode→county fallback is real). FPL figures are **imported** from `FplCache.default["2025"]` rather than copied into the test files, so they cannot drift from the table production reads. That attribute is a plain offline literal — `FplCache.update()` short-circuits with `return self.default` — so there's no network dependency and no flake. `fps` goes one step further and calls the real `FederalPoveryLimit.get_limit` on an unsaved instance instead of reimplementing its arithmetic.

- **Known limitations:** `universal_preschool/tests/test_universal_preschool.py` is moved but not rewritten — it's DB-backed rather than mock-based, unlike everything else here. I left the bodies alone rather than rewriting 8 working tests; say the word if you'd rather it match.
- **Future considerations:** the `tests/test_<program>.py` layout is now consistent across all ten. A few older CO programs still use a flat `tests.py` (e.g. `jeffco_student_benefits`) — worth aligning opportunistically, out of scope here. It's also worth considering whether `Cache` should surface fetch failures instead of silently serving defaults, given point 2.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive coverage for Colorado benefit calculators, including eligibility rules, income thresholds, household and member requirements, benefit values, location criteria, and dependency handling.
  * Expanded boundary and error-case testing for Medicaid, preschool, health care, energy assistance, rebates, and family planning programs.
  * Added coverage for Colorado Universal Preschool eligibility and benefit-hour outcomes across ages, income levels, and qualifying circumstances.
  * Removed obsolete skipped test suites and replaced them with focused calculator-specific coverage.

### CODE REVIEW

**Review by catonph — APPROVED** [review]
Non-blocking: 

1. The FPL constants are hardcoded copies, not the real ones. The module docstrings in all five FPL-using files say "the real 2025 guidelines from FplCache.default", and the PR description says there's no network dependency because FplCache.update() short-circuits. Both are true as facts, but the tests don't actually reach FplCache — they mock program.year.as_dict() and assert against a local FPL_2025 = {...} literal duplicated across five files. 
That's a reasonable unit-test choice and matches wa/head_start, which does the same. But the docstring implies a link that doesn't exist: when 2026 guidelines land, FplCache.default changes and these five copies silently don't. The tests stay green while no longer testing the real bands. Either soften the comment or import the real dict — the latter costs nothing since FplCache.default is a plain offline literal.

2. fps's fpl_limit() helper reimplements production logic. [test_family_planning_services.py](vscode-webview://1jrihtje43b7cumvhk5dugrllr34t814skp8ljhjmlal1lcpccmr/benefits-api/programs/programs/co/medicaid/family_planning_services/tests/test_family_planning_services.py) hand-rolls a fpl_limit() that mirrors [FederalPoveryLimit.get_limit](vscode-webview://1jrihtje43b7cumvhk5dugrllr34t814skp8ljhjmlal1lcpccmr/benefits-api/programs/models.py#L138), including the MAX_DEFINED_SIZE = 8 extrapolation. The docstring says "mirrors ... without touching the database" — but get_limit is pure arithmetic over as_dict() and needs no DB either. Mocking as_dict and calling the real get_limit would test the real function instead of a copy that can drift.

3. Minor: test_age_is_not_a_member_gate doesn't assert what it claims. In [test_omnisalud.py:749](vscode-webview://1jrihtje43b7cumvhk5dugrllr34t814skp8ljhjmlal1lcpccmr/benefits-api/programs/programs/co/omnisalud/tests/test_omnisalud.py#L749), the member is a Mock, so member.age auto-creates regardless — the loop over six ages passes trivially and would keep passing if an age gate were added using any other attribute. It's testing that member_eligible ignores age, which is fine, but the six-value loop implies more coverage than exists.

File: programs/programs/co/medicaid/family_planning_services/tests/test_family_planning_services.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        make_data(medicaid_eligible),
+        Dependencies() if missing_dependencies is None else missing_dependencies,
+    )
+
+
+def eligibility_with_members(count):
+    """An `Eligibility` shaped the way `eligible()` hands it to `household_eligible`."""
+    e = Eligibility()
+    for i in range(count):
+        e.add_member_eligibility(MemberEligibility(make_member()))
```
[the remark this answers is not in the document]
Comment (david-j-stoll) [reply programs/programs/co/medicaid/family_planning_services/tests/test_family_planning_services.py]: Good catch — valid, `i` is genuinely unused in that loop body. Fixed by renaming to `_`, which also matches the existing `for _ in range(2):` a few tests further down in this same file.

Verified with `ruff check --select B007` (clean), `black --check` (clean), and the CO suite (389 passed).

**Comment by david-j-stoll** [comment]
Thanks @catonph — all three were valid, all three are fixed in 8d994b77.

**1. FPL constants were copies, not the real ones.** Correct, and the docstring was asserting a link that nothing enforced. All five files now do `FPL_2025 = FplCache.default["2025"]` instead of a hand-copied literal. As you say it costs nothing — `FplCache.update()` short-circuits with `return self.default`, so it is a plain offline dict with no DB or network access. Verified the imported values are identical to the literals they replaced, so no assertion changed. Docstrings reworded to describe what the code now actually does.

Worth noting the real dict also carries the `"additional"` key that four of my literals had dropped, so the mocked `as_dict()` is now shaped exactly like the real one.

**2. `fpl_limit()` reimplemented production logic.** Agreed, and it was worse than it looked: the `MAX_DEFINED_SIZE` extrapolation branch in that copy had no test covering it, so it was untested duplicate arithmetic. Deleted in favour of an unsaved `FederalPoveryLimit(year="2025", period="2025")` — as you note, `get_limit` is pure arithmetic over `as_dict()` and needs no database. Still wrapped in `Mock(side_effect=...)` so the `assert_called_once_with(6)` checks that pin the sizing bug keep working, but the arithmetic is now the real function.

**3. `test_age_is_not_a_member_gate` did not assert what it claimed.** Right — the Mock auto-creates `.age`, so all six iterations were trivially true. Rather than just trimming the loop I made the claim enforceable: the member is now `Mock(spec=["insurance"])`, so any read of `member.age` raises `AttributeError`. I confirmed this by temporarily adding an age gate to `OmniSalud.member_eligible` — the test fails with `AttributeError: Mock object has no attribute 'age'`, where the old loop would have kept passing.

Full suite green (2986 passed), Black clean, and Ruff clean across all nine authored files.
